File: 169_Majority_Element.py
class Solution(object):
    def majorityElement(self, nums: list) -> int:
        """
        :type nums: List[int]
        :rtype: int
        """
        majority_index = len(nums) / 2
        nums_dict: dict = {}
        for _num in nums:
            if _num in nums_dict.keys():
                nums_dict[_num] += 1
            else:
                nums_dict[_num] = 1
                
        max_key = max(nums_dict, key=lambda k: nums_dict[k] > majority_index)
        return max_key
    
if __name__ == "__main__":
    soln = Solution()
    
    nums = [3,2,3]
    num = soln.majorityElement(nums)
    print(f"majority num : {num}")
    
    nums = [2,2,1,1,1,2,2]
    num = soln.majorityElement(nums)
    print(f"majority num : {num}")

File: 169_Majority_Element.cpp
#include <vector>
#include <iostream>
using namespace std;

class Solution {
public:
    int majorityElement(vector<int>& nums) {
        int n = nums.size();
        unordered_map<int, int> u_m{};

        if (nums.size() == 1) {
            return nums[0];
        }
        
        for (int idx = 0; idx != n; ++idx) {
            if (u_m.find(nums[idx]) == u_m.end()) {
                u_m[nums[idx]] = 1;
            } else {
                u_m[nums[idx]] += 1;
                if (u_m[nums[idx]] > (n / 2)) {
                    return nums[idx];
                }
            }
        }

        return 0;
    }
};

void print_vector(vector<int>& arr) {
    for (auto _ : arr) {
        std::cout << _ << " ";
    }
    putchar('\n');
}

int main() {
    Solution soln = Solution();

    vector<int> nums = { 3,2,3 };
    auto res = soln.majorityElement(nums);
    std::cout << res << std::endl;

    nums = { 2,2,1,1,1,2,2 };
    res = soln.majorityElement(nums);
    std::cout << res << std::endl;

    return 0;
}